![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesores: Diego Cortez, Gabriel Iturra
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Nicolás Cabello, Cristopher Urbina

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1:
- Estudiante n°2:

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

### **Instrucciones importantes**

1. **Formato del informe**:  
   - El informe debe estar integrado dentro de un **Jupyter Notebook**. No es necesario subirlo a una plataforma externa, pero debe cumplir con los siguientes requisitos:  
     - Estructura clara y ordenada.  
     - Código acompañado de explicaciones detalladas.  
     - Resultados presentados de forma visual y analítica.  

2. **Descuento por informes deficientes**:  
   - Cualquier sección del informe que no tenga una explicación adecuada o no respete el formato será penalizada con un descuento en la nota. Esto incluye código sin comentarios o análisis que no sean coherentes con los resultados presentados.
   - **Comentarios sin formatear de ChatGPT o herramientas similares serán penalizados (e.g: "Inserta tu modelo acá", etc.)**

<center>
<img src="https://media1.giphy.com/media/ZkIUY6LZJMwrSMQMWu/giphy.gif?cid=6c09b952msc755426bpsadn4sysofcpqczczi7x1pghd41sw&ep=v1_internal_gif_by_id&rid=giphy.gif&ct=v" width="400" height="300">

# 📬 Entrega Parcial 3 (30% del Proyecto)

### 📪 Fecha de Entrega: 

Esta entrega cuenta con 2 partes:

- **Predicciones**: 1 de Diciembre, 2 de Diciembre, 3 de Diciembre, 4 de Diciembre
- **Informe:** 8 de Diciembre.

## 📖 Enunciado

Después de haber superado con éxito la implementación de sus flujos de trabajo con `Airflow`, el equipo **Deep Drinkers 🤖** ha demostrado que no solo sabe construir modelos, sino también integrarlos en pipelines reales listos para producción.

El equipo técnico de **SodAI Drinks 🥤** ha quedado tan satisfecho con la automatización de sus procesos que ahora les ha encomendado una nueva, y final, misión:  
- 📊 **Evaluar la calidad real de sus modelos en un entorno más desafiante** 
- 🧠 **Comunicar los hallazgos de forma clara y profesional frente a una comisión**.

<center>
<img src="https://i.giphy.com/btbUGSHh3f6eBjbDfh.webp" width="350">


## 📌 Predicciones
<center>
<img src="https://i.giphy.com/NTMxntb8rMzmbd1x97.webp" width="500" height="300">
</center>

Su *pipeline* automatizado debe demostrar que puede manejar datos reales en producción. **SodAI Drinks** 🥤 liberará *batches* de datos en fechas diferentes para evaluar si su sistema se adapta correctamente, los cuales simularán la llegada de información nueva (cada batch es una semana).

Lo que debe funcionar automáticamente:
- **Ejecución Automatizada**: Su DAG de `Airflow` debe ser capaz de procesar cada *batch* de datos de manera automática cuando esté disponible.
- **Monitoreo Continuo**: El *pipeline* debe evaluar la calidad de los datos entrantes y detectar posibles desviaciones respecto al conjunto original.
- **Decisión de Reentrenamiento**: Basándose en los análisis de *data drift* y métricas de rendimiento, el sistema debe determinar automáticamente si es necesario reentrenar el modelo.
- **Tracking Completo**: Todas las ejecuciones, métricas y decisiones deben quedar registradas en `MLFlow` para posterior análisis.
- **Generación de Predicciones**: Para cada *batch* procesado, el sistema debe generar las predicciones correspondientes y almacenarlas de manera organizada.

**IMPORTANTE: Es imperativo que guarden los resultados de su pipeline (métricas, tracking, predicciones, etc) para su análisis en la siguiente sección.**

### 📌Entregable: Competencia en CodaLab

<center>
<img src="https://i.giphy.com/NomCzPIGoXs3EIq77v.webp" width="300" height="300">
</center>

Como parte de esta entrega final, el equipo debe utilizar su pipeline entrenado para **generar predicciones en los nuevos conjuntos de datos mencionados anteriormente**.

Estas predicciones deben:

- Ser generadas directamente desde el *pipeline* previamente desarrollado.  
- Guardarse en archivos `.csv` siguiendo el formato requerido.  
- Subirse a la plataforma **CodaLab**, donde se realizará la evaluación final.

En CodaLab podrán:

- Ver el rendimiento de su modelo frente a los de otros equipos.  
- Obtener una puntuación basada en la métrica definida del proyecto.  

Cada set de predicciones debe ser publicado en CodaLab en las siguientes fechas:
- **Predicciones 1 📅** : 1 de Diciembre (Fecha de predicción: 01/01/25 - 05/01/25)
    - Note que para este set de predicciones, sólo tendrán acceso a los datos publicados en la Entrega 1.
- **Predicciones 2 📅** : 2 de Diciembre (Fecha de predicción: 06/01/25 - 12/01/25)
- **Predicciones 3 📅** : 3 de Diciembre (Fecha de predicción: 13/01/25 - 19/01/25)
- **Predicciones 4 📅** : 4 de Diciembre (Fecha de predicción: 20/01/25 - 26/01/25)

Para simular la llegada de nuevos datos, se publicarán los siguientes conjuntos de datos:
- **Batch 1 📅** : 2 de Diciembre (Con datos realizados del período 01/01/25 - 05/01/25)
- **Batch 2 📅** : 3 de Diciembre (Con datos realizados del período 06/01/25 - 12/01/25)
- **Batch 3 📅** : 4 de Diciembre (Con datos realizados del período 13/01/25 - 19/01/25)
- **Batch 4 📅** : 5 de Diciembre (Con datos realizados del período 20/01/25 - 26/01/25)
    - Note como este conjunto de datos se publica **después** de la competencia, por lo que sólo les servirá para la última evaluación requerida en el informe.

**IMPORTANTE:** Subir los resultados a tiempo en las tres fechas es **<u>OBLIGATORIO</u>** para la evaluación del desempeño final del equipo. **Por cada fecha en la que no se suban predicciones, se aplicará un descuento de 0.75 puntos (75 décimas) sobre la nota de la Entrega 3.**

### 🎁 Bonus [0.5 puntos]

Como incentivo adicional, se premiará a los **3 equipos con mejor performance en la métrica F1** con **0.5 puntos de puntaje adicional** sobre la nota de la Entrega 3.

## 📌 Informe: Performance del modelo [6.0 puntos]

En esta sección se espera que el equipo analice de manera retrospectiva el performance de su modelo y saque conclusiones en función de sus resultados. 

Para esto, tendrán que incurrir en 3 tipos de análisis:

- **Individual:** Deben evaluar el performance de su modelo tomando cada semana de forma aislada.

- **Comparativo:** Deben evaluar el performance de su modelo *a través* de las semanas, comparando el desempeño entre las semanas e identificando posibles tendencias.
    - Para esta parte se espera que generen gráficos de tendencia y tablas comparativas para apoyar su análisis.

- **Conclusiones y Aprendizajes:** Deben escribir sus principales conclusiones y aprendizajes de este proyecto. 

A lo largo de esta sección, se espera que respondan preguntas como:

1. ¿Cómo variaron sus métricas a lo largo de los distintos conjuntos de datos? 
2. ¿En qué momento el modelo tuvo su peor desempeño y por qué?
3. ¿Detectaron algún cambio significativo (drift) en la distribución de los datos? ¿Cómo lo identificaron?  
4. ¿Tuvieron que reentrenar su modelo con los nuevos datos? ¿Porqué? ¿Ayudó esto al performance de su modelo?
4. ¿Qué decisión técnica (modelo, métrica, imputación, etc.) tuvo más impacto en los resultados?  
5. ¿Qué hiperparámetro fue el más importante para su modelo? 
5. ¿Qué variable fue más influyente en las predicciones? ¿Cómo lo interpretan? ¿Cómo cambió su importancia con respecto a las otras variables a lo largo de los batch de datos?
6. ¿Qué aprendieron sobre el negocio a partir de los resultados del modelo?
7. ¿Qué limitaciones detectaron en su modelo o en los datos?  

**IMPORTANTE: Se espera que en sus respuestas hagan referencia a los artefactos (métricas, hiperparámetros, gráficos de interpretabilidad, etc) que su pipeline genera.**

<center>
<img src="https://media1.tenor.com/m/jyQ3SPT1htEAAAAd/i-love-this-performance-even-more-simon-cowell.gif" width="300" height="300">
</center>

### 📌 Análisis individual [3.0 puntos]

In [28]:
# Importamos las librerías necesarias
import pandas as pd
import numpy as np
from pathlib import Path
from sklearn.metrics import classification_report

# Funciones auxiliares para leer datos y calcular métricas

# Leer las predicciones desde un archivo CSV
def _read_predictions(pred_path: Path) -> pd.DataFrame:
    try:
        df = pd.read_csv(pred_path, header=None, names=["customer_id", "product_id"])
    except Exception:
        df = pd.read_csv(pred_path)
        df = df[["customer_id", "product_id"]]
    return df.drop_duplicates()

# Leer las referencias de verdad desde un archivo Parquet o CSV
def _read_reference(truth_path: Path) -> pd.DataFrame:
    if truth_path is None:
        return None
    truth = None
    try:
        truth = pd.read_parquet(truth_path)
    except Exception:
        alt_csv = Path(truth_path).with_suffix(".csv")
        if alt_csv.exists():
            truth = pd.read_csv(alt_csv)
        else:
            raise
    truth = truth[["customer_id", "product_id"]].drop_duplicates()
    return truth

# Calcular métricas de precisión, recall y F1 basadas en conjuntos
def _compute_set_f1(preds: pd.DataFrame, truth: pd.DataFrame):
    pred_pairs = set(map(tuple, preds[["customer_id", "product_id"]].to_numpy()))
    truth_pairs = set(map(tuple, truth[["customer_id", "product_id"]].to_numpy()))
    tp = len(pred_pairs & truth_pairs)
    fp = len(pred_pairs - truth_pairs)
    fn = len(truth_pairs - pred_pairs)
    precision = tp / (tp + fp + 1e-9)
    recall = tp / (tp + fn + 1e-9)
    f1 = 2 * precision * recall / (precision + recall + 1e-9)
    return tp, fp, fn, precision, recall, f1, len(pred_pairs), len(truth_pairs)

# Preparar los vectores y_true y y_pred para el informe de clasificación
def _ytrue_ypred_for_report(preds: pd.DataFrame, truth: pd.DataFrame):
    pred_pairs = set(map(tuple, preds[["customer_id", "product_id"]].to_numpy()))
    truth_pairs = set(map(tuple, truth[["customer_id", "product_id"]].to_numpy()))
    universe = sorted(pred_pairs | truth_pairs)
    y_true = np.array([1 if pair in truth_pairs else 0 for pair in universe])
    y_pred = np.array([1 if pair in pred_pairs else 0 for pair in universe])
    return y_true, y_pred

# Función principal para evaluar las predicciones
def eval_batch_predictions(pred_path: Path, truth_parquet: Path | None = None, with_report: bool = True):
    preds = _read_predictions(pred_path)
    if truth_parquet and (Path(truth_parquet).exists() or Path(truth_parquet).with_suffix(".csv").exists()):
        truth = _read_reference(truth_parquet)
        tp, fp, fn, p, r, f1, n_pred, n_truth = _compute_set_f1(preds, truth)
        print(f"Pares de predicción únicos: {n_pred} | Pares de verdad únicos: {n_truth}")
        print(f"Verdaderos Positivos (TP): {tp} | Falsos Positivos (FP): {fp} | Falsos Negativos (FN): {fn}")
        print(f"Precisión: {p:.6f} | Recall: {r:.6f} | F1: {f1:.6f}")
        if with_report:
            y_true, y_pred = _ytrue_ypred_for_report(preds, truth)
            print("\nClassification Report (positivo=1, negativo=0):")
            print(classification_report(y_true, y_pred, digits=6))
    else:
        counts = preds.groupby("customer_id").size()
        print(f"Filas: {len(preds)} | Clientes: {counts.size} | top-k medio: {counts.mean():.2f} | top-k máximo: {counts.max()}")

#### 📌 Batch 1 (01/01/25 - 05/01/25) [0.75 puntos]

In [29]:
# Evaluación de Batch 1
pred_batch1 = Path("predictions/predicciones_4.csv")
truth_batch1 = Path("data/batch_t1.parquet")
if pred_batch1.exists():
    eval_batch_predictions(pred_batch1, truth_batch1, with_report=True)
else:
    print("Archivo de Batch 1 no encontrado:", pred_batch1)

Pares de predicción únicos: 5628 | Pares de verdad únicos: 3264
Verdaderos Positivos (TP): 1993 | Falsos Positivos (FP): 3635 | Falsos Negativos (FN): 1271
Precisión: 0.354122 | Recall: 0.610600 | F1: 0.448268

Classification Report (positivo=1, negativo=0):
              precision    recall  f1-score   support

           0   0.000000  0.000000  0.000000      3635
           1   0.354122  0.610600  0.448268      3264

    accuracy                       0.288882      6899
   macro avg   0.177061  0.305300  0.224134      6899
weighted avg   0.167540  0.288882  0.212081      6899



Interpretación de resultados:

- Cobertura: Las predicciones incluyen 5,628 pares únicos frente a 3,264 pares reales.
- Desempeño: F1 = 0.448; precisión moderada (0.354) y recall alto (0.611).
- Lectura: El modelo privilegia recall (recupera muchos verdaderos) a costa de falsos positivos. Esto sugiere un umbral relativamente bajo o una estrategia de ranking con top-k amplio.
- Implicancias: Útil para no perder compras reales, pero incrementa carga operativa por recomendaciones irrelevantes. Ajustar top-k o mezcla de señales puede mejorar precisión sin sacrificar demasiado recall.
- Próximo ajuste: bajar un poco el top-k o subir el umbral para mejorar precisión, cuidando no perder demasiado recall.

#### 📌 Batch 2 (06/01/25 - 12/01/25) [0.75 puntos]

In [30]:
# Evaluación de performance por batch (Batch 2)
import pandas as pd
import numpy as np
from pathlib import Path

# Batch 2 → Mejor predicción (CSV)
pred_batch2 = Path("predictions/predicciones_batch_t1_adv_b.csv")
truth_batch2 = Path("data/batch_t2.parquet")
if pred_batch2.exists():
    eval_batch_predictions(pred_batch2, truth_batch2, with_report=True)
else:
    print("Archivo de Batch 2 no encontrado:", pred_batch2)

Pares de predicción únicos: 2918 | Pares de verdad únicos: 5103
Verdaderos Positivos (TP): 1630 | Falsos Positivos (FP): 1288 | Falsos Negativos (FN): 3473
Precisión: 0.558602 | Recall: 0.319420 | F1: 0.406433

Classification Report (positivo=1, negativo=0):
              precision    recall  f1-score   support

           0   0.000000  0.000000  0.000000      1288
           1   0.558602  0.319420  0.406433      5103

    accuracy                       0.255046      6391
   macro avg   0.279301  0.159710  0.203217      6391
weighted avg   0.446025  0.255046  0.324523      6391



Interpretación de resultados:

- Cobertura: 2,918 pares predichos vs 5,103 pares reales.
- Desempeño: F1 = 0.406; precisión 0.559 y recall 0.319.
- Lectura: Mayor precisión que en Batch 1, pero recall disminuye. Básicamente ahora acertamos más proporcionalmente (mayor precisión), pero dejamos pasar más compras reales (menor recall). El sistema está más conservador: recomienda menos y acierta más proporcionalmente.
- Implicancias: Buen control de falsos positivos; podría recuperar más verdaderos ampliando top-k o agregando señales de popularidad/recencia para explorar sin perder demasiado precisión.
- Próximo ajuste: subir un poco el top-k o bajar levemente el umbral para recuperar recall, manteniendo una precisión razonable.

Interpretación de resultados:

- Cobertura: 2,918 pares predichos vs 5,103 pares reales.
- Desempeño: F1 = 0.406; precisión 0.559 y recall 0.319.
- Lectura: Mayor precisión que en Batch 1, pero recall disminuye. El sistema está más conservador: recomienda menos y acierta más proporcionalmente.
- Implicancias: Buen control de falsos positivos; podría recuperar más verdaderos ampliando top-k o agregando señales de popularidad/recencia para explorar sin perder demasiado precisión.

#### 📌 Batch 3 (13/01/25 - 19/01/25) [0.75 puntos]

In [31]:
# Evaluación de performance por batch (Batch 3)
import pandas as pd
import numpy as np
from pathlib import Path

# Batch 3 → Mejor predicción (ZIP o CSV)
pred_batch3 = Path("predictions/predicciones_batch_t2_adv_b.zip")
truth_batch3 = Path("data/batch_t3.parquet")
if pred_batch3.exists():
    eval_batch_predictions(pred_batch3, truth_batch3, with_report=True)
else:
    alt_csv_3 = Path("predictions/predicciones_batch_t2_adv_b.csv")
    if alt_csv_3.exists():
        eval_batch_predictions(alt_csv_3, truth_batch3, with_report=True)
    else:
        print("Archivo de Batch 3 no encontrado:", pred_batch3)

Pares de predicción únicos: 4235 | Pares de verdad únicos: 4922
Verdaderos Positivos (TP): 2199 | Falsos Positivos (FP): 2036 | Falsos Negativos (FN): 2723
Precisión: 0.519244 | Recall: 0.446770 | F1: 0.480288

Classification Report (positivo=1, negativo=0):
              precision    recall  f1-score   support

           0   0.000000  0.000000  0.000000      2036
           1   0.519244  0.446770  0.480288      4922

    accuracy                       0.316039      6958
   macro avg   0.259622  0.223385  0.240144      6958
weighted avg   0.367307  0.316039  0.339750      6958



**Interpretación de resultados:**

- Cobertura: 4,235 pares predichos vs 4,922 reales.
- Desempeño: F1 = 0.480; precisión 0.519 y recall 0.447.
- Lectura: Mejor equilibrio entre precisión y recall respecto a los anteriores. Acierta más y recupera una proporción mayor de verdaderos, es decir, se logró un balance mejor entre acertar y recuperar compras reales.
- Implicancias: La combinación actual (estrategia `adv_b`/mejoras recientes) parece estabilizar el trade-off. Se puede afinar con blending de popularidad o ajuste de umbral por cliente para ganar puntos de F1 adicionales.
- Próximo ajuste: afinar top-k por cliente según probabilidad/score y revisar umbrales por segmento para sumar algunos puntos de F1.

#### 📌 Batch 4 (20/01/25 - 26/01/25) [0.75 puntos]

In [32]:
# Evaluación de performance por batch (Batch 4)
import pandas as pd
import numpy as np
from pathlib import Path

# Batch 4 → Mejor predicción (ZIP o CSV)
pred_batch4 = Path("predictions/predicciones_batch_t3_adv_hybrid2.zip")
truth_batch4 = Path("data/batch_t4.parquet")
if pred_batch4.exists():
    eval_batch_predictions(pred_batch4, truth_batch4)
else:
    alt_csv_4 = Path("predictions/predicciones_batch_t3_adv_hybrid2.csv")
    if alt_csv_4.exists():
        eval_batch_predictions(alt_csv_4, truth_batch4)
    else:
        print("Archivo de Batch 4 no encontrado:", pred_batch4)

Filas: 4283 | Clientes: 1012 | top-k medio: 4.23 | top-k máximo: 20


### 📌 Análisis comparativo [2.0 puntos]

In [33]:
# código para realizar análisis de resultados entre batchs

# Análisis comparativo: consolidación y tendencias
import pandas as pd
import numpy as np
import plotly.express as px
from pathlib import Path

# Reutilizamos las funciones definidas anteriormente: _read_predictions, _read_reference, _compute_set_f1

def evaluate_batch_metrics(name: str, pred_path: Path, truth_path: Path | None):
    if not pred_path.exists():
        return {"batch": name, "precision": np.nan, "recall": np.nan, "f1": np.nan, "tp": np.nan, "fp": np.nan, "fn": np.nan, "n_pred": np.nan, "n_truth": np.nan}
    if truth_path is None or (not Path(truth_path).exists() and not Path(truth_path).with_suffix(".csv").exists()):
        # Sin ground truth aún → solo cobertura
        preds = _read_predictions(pred_path)
        counts = preds.groupby("customer_id").size()
        return {"batch": name, "precision": np.nan, "recall": np.nan, "f1": np.nan, "tp": np.nan, "fp": np.nan, "fn": np.nan, "n_pred": len(preds.drop_duplicates()), "n_truth": np.nan, "clientes": counts.size, "topk_mean": counts.mean(), "topk_max": counts.max()}
    # Con ground truth → métricas completas
    preds = _read_predictions(pred_path)
    truth = _read_reference(truth_path)
    tp, fp, fn, p, r, f1, n_pred, n_truth = _compute_set_f1(preds, truth)
    return {"batch": name, "precision": p, "recall": r, "f1": f1, "tp": tp, "fp": fp, "fn": fn, "n_pred": n_pred, "n_truth": n_truth}

# Paths de predicciones y truths
pred_paths = {
    "Batch 1": Path("predictions/predicciones_4.csv"),
    "Batch 2": Path("predictions/predicciones_batch_t1_adv_b.csv"),
    "Batch 3": Path("predictions/predicciones_batch_t2_adv_b.zip"),
    "Batch 4": Path("epredictions/predicciones_batch_t3_adv_hybrid2.zip"),
}
truth_paths = {
    "Batch 1": Path("data/batch_t1.parquet"),
    "Batch 2": Path("data/batch_t2.parquet"),
    "Batch 3": Path("data/batch_t3.parquet"),
    "Batch 4": Path("data/batch_t4.parquet") if Path("entrega_3/data/batch_t4.parquet").exists() else None,
}

rows = []
for name in ["Batch 1", "Batch 2", "Batch 3", "Batch 4"]:
    rows.append(evaluate_batch_metrics(name, pred_paths[name], truth_paths[name]))

metrics_df = pd.DataFrame(rows)
print("Resumen de métricas por batch:")
print(metrics_df)

# Gráfico de tendencias (se actualiza cuando Batch 4 tenga truth)
plot_df = metrics_df.copy()
plot_df = plot_df.dropna(subset=["f1"])  # solo batches con F1 disponible
fig = px.line(plot_df, x="batch", y=["precision", "recall", "f1"], markers=True,
              title="Tendencia de métricas por batch")
fig.update_layout(legend_title_text="Métrica")
fig.show()

Resumen de métricas por batch:
     batch  precision   recall        f1      tp      fp      fn  n_pred  \
0  Batch 1   0.354122  0.61060  0.448268  1993.0  3635.0  1271.0  5628.0   
1  Batch 2   0.558602  0.31942  0.406433  1630.0  1288.0  3473.0  2918.0   
2  Batch 3   0.519244  0.44677  0.480288  2199.0  2036.0  2723.0  4235.0   
3  Batch 4        NaN      NaN       NaN     NaN     NaN     NaN     NaN   

   n_truth  
0   3264.0  
1   5103.0  
2   4922.0  
3      NaN  


### 📌 Conclusiones y Aprendizajes [1.0 puntos]

> [Escriba aquí sus conclusiones y aprendizajes]

## 📁 Estructura de entrega

Se espera que su entrega cuente con la siguiente estructura de carpetas:

```bash
entrega_3/
│
├── informe.ipynb # Jupyter notebook con sus resultados
│
├── predictions/ # Carpeta con las predicciones para cada batch de datos
│
└── data/ # Carpeta con los datos realizados (lo que realmente ocurrió)
```

Mucho éxito!

<center>
<img src="https://gifdb.com/images/high/may-the-force-be-with-you-anakin-skywalker-n2g5o0pm4h6iylsx.gif" width="400" height="300">
